snuba/query/processors/tagsmap.py

```diff
@@ -0,0 +1,125 @@
+from snuba.util import is_condition, is_function
+
+
+TAG_PATTERN = re.compile(r"([a-zA-Z0-9_\.]+)\[([a-zA-Z0-9_\.]+)\]")
+
+
+class Operand(Enum):
+    EQ = "="
+    NEQ = "!="
+
+
+class OptimizableCondition(NamedTuple):
+    nested_col_index: str
+    operand: Operand
+    value: str
+
+
+class NestedFieldConditionOptimizer(QueryProcessor):
+    """
+    This processor scans the the conditions in the query and converts
+    top level eq/neq conditions on specific nested fields, into LIKE/NOT LIKE
+    statements on a column that merges all the content of a nested one.
+
+    Example:
+    tags: {
+        "tag1": "1",
+        "tag2": "2",
+    }
+
+    The merged columns looks like:
+    tags_map: "tag1:1|tag2:2"
+
+    This means that a simple EQ/NEQ condition like `tags[tag1] == 1` can be
+    transformed into `tags_map LIKE "%tag1:2%"`
+    """
+
+    def __init__(self, nested_col: str, merged_col: str) -> None:
+        self.__nested_col = nested_col
+        self.__merged_col = merged_col
+
+    def __is_optimizable(
+        self, condition: Condition, column: str
+    ) -> Optional[OptimizableCondition]:
+        """
+        Recognize if the condition can be optimized.
+        This includes these kind of conditions:
+        - top level conditions. No nested OR
+        - the condition has to be either in the form tag[t] = value
+        - functions referencing the tags as parameters are not taken
+          into account except for ifNull.
+        - Both EQ and NEQ conditions are optimized.
+        """
+        if not is_condition(condition):
+            return None
+        if condition[1] not in ["=", "!="]:
+            return None
+        if not isinstance(condition[2], str):
+            # We can only support literals for now.
+            return None
+        lhr = condition[0]
+
+        # This unpoacks the ifNull function. This is just an optimization to make this class more
+        # useful since the product wraps tags access into ifNull very often and it is a trivial
+        # function to unpack. We could exptend it to more functions later.
+        function_expr = is_function(lhr, 0)
+        if function_expr and function_expr[0] == "ifNull" and len(function_expr[1]) > 0:
+            lhr = function_expr[1][0]
+        if not isinstance(lhr, str):
+            return None
+
+        # Now we have a condition in the form of: ["tags[something]", "=", "a string"]
+        tag = TAG_PATTERN.match(lhr)
+        if tag and tag[1] == self.__nested_col:
+            # tag[0] is the full expression that matches the re.
+            nested_col_index = tag[2]
+            return OptimizableCondition(
+                nested_col_index=nested_col_index,
+                operand=Operand.EQ if condition[1] == "=" else Operand.NEQ,
+                value=condition[2],
+            )
+        return None
+
+    def process_query(self, query: Query, request_settings: RequestSettings,) -> None:
```

Comment:
```suggestion
    def process_query(self, query: Query, request_settings: RequestSettings) -> None:
```

I don't understand why `black` doesn't fix these. 😕

Reply:
it is worse. If you write (like I did)
```
def process_query(
    self,
    query,
):
```
Which is correct. It will make it wrong by turning it into:
```
def process_query(self, query,)
```

by preserving the comma at the end. I suspect it is not the intended behavior, and that, if I did not add the final comma in the first place it would have been rewritten without.